File: Python3/846. Hand of Straights.py
class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        if len(hand) % groupSize:
            return False

        cards = Counter(hand)
        values = list(cards.keys())
        heapq.heapify(values)

        while values:
            cur = values[0]
            for i in range(cur, cur + groupSize):
                if i not in cards:
                    return False
                cards[i] -= 1

                if cards[i] == 0:
                    del cards[i]
                    if i != heapq.heappop(values):
                        return False

        return True

File: Java/846. Hand of Straights.java
class Solution {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        if (hand.length % groupSize != 0) {
            return false;
        }

        Map<Integer, Integer> cards = new HashMap<>();
        for (int h: hand) {
            cards.put(h, cards.getOrDefault(h, 0) + 1);
        }

        for (int h: hand) {
            int start_card = h;
            // Find the start of this chain of sequence
            while (cards.getOrDefault(start_card - 1, 0) != 0) {
                start_card--;
            }

            // Process the sequence starting from start_card
            while (start_card <= h) {
                while (cards.get(start_card) > 0) {
                    for (int next_card = start_card; next_card < start_card + groupSize; next_card++) {
                        if (cards.getOrDefault(next_card, 0) == 0) {
                            return false;
                        }
                        cards.put(next_card, cards.get(next_card) - 1);
                    }
                }
                start_card++;
            }
        }
        return true;
    }
}